lib/Extractor/ExprBuilder.cpp

```diff
@@ -0,0 +1,945 @@
+// Copyright 2014 The Souper Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "llvm/Support/CommandLine.h"
+#include "souper/Extractor/ExprBuilder.h"
+
+#include <queue>
+
+using namespace llvm;
+using namespace souper;
+
+namespace souper {
+
+static llvm::cl::opt<souper::ExprBuilder::Builder> SMTExprBuilder(
+    "souper-smt-expr-builder",
+    llvm::cl::Hidden,
+    llvm::cl::desc("SMT-LIBv2 expression builder (default=klee)"),
+    llvm::cl::values(clEnumValN(souper::ExprBuilder::KLEE, "klee",
+                                "Use KLEE's Expr library")),
+    llvm::cl::init(souper::ExprBuilder::KLEE));
+
+bool ExprBuilder::getUBPaths(Inst *I, UBPath *Current,
+                             std::vector<std::unique_ptr<UBPath>> &Paths,
+                             UBPathInstMap &CachedUBPathInsts, unsigned Depth) {
+  if (Depth > MAX_PHI_DEPTH)
+    return false;
+
+  switch (I->K) {
+    default:
+      break;
+
+    case Inst::AddNSW:
+    case Inst::AddNUW:
+    case Inst::AddNW:
+    case Inst::SubNSW:
+    case Inst::SubNUW:
+    case Inst::SubNW:
+    case Inst::MulNSW:
+    case Inst::MulNUW:
+    case Inst::MulNW:
+    case Inst::UDiv:
+    case Inst::SDiv:
+    case Inst::UDivExact:
+    case Inst::SDivExact:
+    case Inst::URem:
+    case Inst::SRem:
+    case Inst::Shl:
+    case Inst::ShlNSW:
+    case Inst::ShlNUW:
+    case Inst::ShlNW:
+    case Inst::LShr:
+    case Inst::LShrExact:
+    case Inst::AShr:
+    case Inst::AShrExact:
+      Current->UBInsts.push_back(I);
+      break;
+  }
+
+  const std::vector<Inst *> &Ops = I->orderedOps();
+  if (I->K == Inst::Phi) {
+    // Early terminate because this phi has been processed.
+    // We will use its cached predicates.
+    if (CachedUBPathInsts.count(I))
+      return true;
+    Current->Insts.push_back(I);
+    // Since we treat a select instruction as a phi instruction, it's
+    // possible that I->B has been added already.
+    if (Current->BlockConstraints.count(I->B))
+      return true;
+    std::vector<UBPath *> Tmp = { Current };
+    // Create copies of the current path
+    for (unsigned J = 1; J < Ops.size(); ++J) {
+      UBPath *New = new UBPath;
+      *New = *Current;
+      New->BlockConstraints[I->B] = J;
+      Paths.push_back(std::move(std::unique_ptr<UBPath>(New)));
```

Comment:
```std::vector``` is move-aware. We can simply use the follow form:

    Paths.push_back(std::unique_ptr<UBPath>(New));

I notice there are several other similar places where we can make the simplification.